zuul-core/src/main/java/com/netflix/zuul/message/http/HttpRequestMessageImpl.java

```diff
@@ -563,6 +557,48 @@ public Optional<Integer> getClientDestinationPort() {
         }
     }
 
+    /**
+     * Attempt to parse the Host header from the collection of headers
+     * and return the hostname and port components.
+     *
+     * @return Hostname and Port pair.
+     *         Hostname may be null. Port may be -1 when no valid port is found in the host header.
+     */
+    private static Pair<String, Integer> parseHostHeader(Headers headers) throws URISyntaxException {
+        String host = headers.getFirst(HttpHeaderNames.HOST);
+        if (host == null) {
+            return new Pair<>(null, -1);
+        }
+
+        try {
+            // attempt to use default URI parsing - this can fail when not strictly following RFC2396,
+            // for example, having underscores in host names will fail parsing
+            URI uri = new URI(/* scheme= */ null, host, /* path= */ null, /* query= */ null, /* fragment= */ null);
+            if (uri.getHost() != null) {
+                return new Pair<>(uri.getHost(), uri.getPort());
+            }
+        } catch (URISyntaxException ignored) { }
+
+        // fallback to using a colon split
+        // valid IPv6 addresses would have been handled already so any colon is safely assumed a port separator
```

Comment:
What if it's an invalid IPv6 address?

Reply:
There is a fallback below that detects more than 1 colon (e.g. a non-bracketed IPv6 address) which will throw an exception